src/table/system.rs

```diff
@@ -71,6 +73,32 @@ impl<View: SystemTableView> SystemTable<View> {
     pub fn config_table(&self) -> &[cfg::ConfigTableEntry] {
         unsafe { slice::from_raw_parts(self.table.cfg_table, self.table.nr_cfg) }
     }
+
+    /// Creates a new `SystemTable<View>` from a raw address. The address might
+    /// come from the Multiboot2 information structure or something similar.
+    ///
+    /// # Example
+    /// ```rust
+    /// use core::ffi::c_void;
+    /// use uefi::prelude::{Boot, SystemTable};
+    ///
+    /// let system_table_addr = 0xdeadbeef as *mut c_void;
+    ///
+    /// let mut uefi_system_table = unsafe {
+    ///     SystemTable::<Boot>::from_ptr(system_table_addr).expect("Pointer must not be null!")
+    /// };
+    /// ```
+    ///
+    /// # Safety
+    /// This function is unsafe, because the caller must be sure, that the pointer
+    /// is valid. Otherwise, further operations on the object might result in
+    /// undefined behaviour, even if the methods might not be marked as unsafe.
+    pub unsafe fn from_ptr(ptr: *mut c_void) -> Option<Self> {
+        // This is safe, because `Self` has a transparent representation, contains a
+        // reference to "SystemTableImpl", and that reference is a pointer in memory
+        // to the actual table layout.
+        NonNull::new(ptr).map(|ptr| core::mem::transmute(ptr.as_ptr()))
```

Comment:
I think the transmute can also be avoided for this one, something like:

```rust
pub unsafe fn from_ptr(ptr: *mut c_void) -> Option<Self> {
        if ptr.is_null() {
            None
        } else {
            Some(Self {
                table: &*(ptr as *mut _),
                _marker: PhantomData,
            })
        }
    }
```

Didn't pull in `NonNull` here since unlike `Handle`, `SystemTable` doesn't use a `NonNull` pointer internally, and don't want to accidentally run afoul of the warning in the [docs](https://doc.rust-lang.org/std/ptr/struct.NonNull.html): "This is often the correct thing to use when building data structures using raw pointers, but is ultimately more dangerous to use because of its additional properties. If you’re not sure if you should use NonNull<T>, just use *mut T!".

Reply:
Can you further explain, why we should not use `NonNull`? In my understanding, `NonNull` ensures that the pointer is not null, because it returns `None`, if the pointer is null. Isn't this what we want here? A raw pointer that is definitely not null? @nicholasbishop

Reply:
Here's my thought process re. the `SystemTable`:

1. As you said, we want a raw pointer that isn't null.
2. We could express that as an explicit `ptr.is_null()` check as in my previous comment, or using the `NonNull::new` constructor.
3. IMO neither one strongly wins on readability alone.
4. The [NonNull](https://doc.rust-lang.org/std/ptr/struct.NonNull.html) documentation says that `NonNull` has additional restrictions beyond those attached to `*mut`.

Now, I don't see anything in the documentation of `NonNull` that makes me think this specific use is not valid; I'm just not an expert in Rust unsafeness so I find it hard to be sure.

All that said, I think the main thing I care about changing here is to avoid `transmute`, so if you prefer to keep the `NonNull` I'd be OK with this too:
```
NonNull::new(ptr as *mut SystemTableImpl).map(|ptr| Self {
    table: ptr.as_ref(),
    _marker: PhantomData,
})
```

Hope that clarifies it :)

Reply:
I'd stick with 'NonNull` because this is consistent with the existing code inside the crate. However, yeah, a good idea to get rid of the transmute.

I updated the PR @nicholasbishop